Fix order of hook execution

thiolliere opened the pull request:
Fix https://github.com/paritytech/substrate/issues/6280
Fix https://github.com/paritytech/substrate/issues/9105

polkadot companion: https://github.com/paritytech/polkadot/pull/4181
cumulus companion: https://github.com/paritytech/cumulus/pull/711

# Breaking Change:

* `AllPallets` is deprecated in favor of explicit: `AllPalletsWithSystem`, `AllPalletsWithoutSystem`, `AllPalletsWithSystemReversed`, `AllPalletsWithoutSystemReversed`. Before this PR `AllPallets` represented all pallets without system in reversed order (i.e. same as `AllPalletsWithoutSystemReversed`, it now represent `AllPalletsWithSystem` as it is a more sane expectation.
* `frame_executive::Executive` requires as generic all pallets **with** system. Before this PR the type required as generic all pallets **without** system and always executed system pallet hook before all pallets hook. But this requires duplicate works and error prone implementation thus it now requires all pallets **with** system. The previous order was to execute the hook on `system` pallet first and then `AllPallets`. It is recomended to use one of `AllPalletsWithSystem` or `AllPalletsWithSystemReversed` or `AllPalletsReversedWithSystemFirst`.

Because before the PR all pallets represented all pallets in reversed, and system was executed first then if someone want to keep the same behavior as before they should use the type `AllPalletsReversedWithSystemFirst` when using the type `frame_executive::Executive`.
Though it is recommended to be careful about the order of definition of the pallet in construct_runtime. and probably reorganize and use `AllPalletsWithSystem`, so the order is the expected normal order and not in reversed (as before this PR).

NOTE: frame_executive::Executive will execute the hooks `on_initialize`, `on_idle`, `on_runtime_upgrade`, `on_finalize`, `offchain_worker` on the generic `AllPallets`, so the order should be reviewed for those execution.


# Upgrade guide:

1. use `AllPalletsReversedWithSystemFirst` when instancing executive type, that mean the execution order is not changed from before this PR, thus no need to change anything in `construct_runtime`.

2. Use the new `AllPalletsWithSystem`, but this would mean that they might need to reorder some stuff in `construct_runtime`.

3. review other usage of `AllPallets` and change to `AllPalletsWithoutSystemReversed` in case the old order must be kept, or change to `AllPalletsWithoutSystem` to get the new definition with pallets defined in the order of their declaration in `construct_runtime`.

thiolliere commented:
for the polkadot It seems we don't need any changes:
maybe scheduler should be put last instead of 2nd. So that the execution of scheduled call happens after other pallet initialization ?

shawntabrizi commented on frame/executive/src/lib.rs:
```diff
@@ -148,7 +148,8 @@ pub type OriginOf<E, C> = <CallOf<E, C> as Dispatchable>::Origin;
 /// - `Context`: The context that is used when checking an extrinsic.
 /// - `UnsignedValidator`: The unsigned transaction validator of the runtime.
-/// - `AllPallets`: Tuple that contains all modules. Will be used to call e.g. `on_initialize`.
+/// - `AllPallets`: Tuple that contains all pallets including frame system pallet. Will be used to
```
```suggestion
/// - `AllPalletsWithSystem`: Tuple that contains all pallets including frame system pallet. Will be used to
```

thiolliere replied:
fixed in https://github.com/paritytech/substrate/pull/10043/commits/b41e9359a4a156fdf42c91db031ef25456081278

shawntabrizi commented on frame/executive/src/lib.rs:
```diff
@@ -148,7 +148,8 @@ pub type OriginOf<E, C> = <CallOf<E, C> as Dispatchable>::Origin;
+///   call hooks e.g. `on_initialize`.
 /// - `OnRuntimeUpgrade`: Custom logic that should be called after a runtime upgrade. Modules are
 ///   already called by `AllPallets`. It will be called before all modules will be called.
 pub struct Executive<System, Block, Context, UnsignedValidator, AllPallets, OnRuntimeUpgrade = ()>(
```
```suggestion
pub struct Executive<System, Block, Context, UnsignedValidator, AllPalletsWithSystem, OnRuntimeUpgrade = ()>(
```

thiolliere replied:
fixed in https://github.com/paritytech/substrate/pull/10043/commits/b41e9359a4a156fdf42c91db031ef25456081278

shawntabrizi commented on frame/executive/src/lib.rs:
```diff
@@ -256,19 +250,10 @@ where
-			OnRuntimeUpgrade
-		>::pre_upgrade().unwrap();
-
+		<(COnRuntimeUpgrade, AllPallets) as OnRuntimeUpgrade>::pre_upgrade().unwrap();
```
I think this is also a noteworthy change. Here system pallet upgrades after the runtime upgrades.

Dunno actually if this can cause issues?

thiolliere replied:
I think for pre_upgrade/post_upgrade the order wasn't so important as it is just checks.
But the order of on_runtime_upgrade is important. That said the order was: custom, then system then pallets.
I will precise this in the note

kianenigma replied:
`try_on_runtime_upgrade` should always imitate what the real on_runtime_upgrade would do, which is respected here, so no issues.

shawntabrizi commented on frame/support/procedural/src/construct_runtime/mod.rs:
```diff
@@ -204,35 +204,64 @@ fn decl_all_pallets<'a>(
+			Note that the type `frame_executive::Executive` expects one of `AllPalletsWithSystem` \
+			and `AllPalletsWithSystemReversed. More details in \
+			https://github.com/paritytech/substrate/pull/10043")]
+		pub type AllPallets = AllPalletsWithSystem;
```
This is wrong right?

If you are going to provide the old `AllPallets` type, it should have the same behavior has before?

shawntabrizi commented on frame/support/procedural/src/construct_runtime/mod.rs:
```diff
@@ -204,35 +204,64 @@ fn decl_all_pallets<'a>(
-		pub type AllModulesWithSystem = ( #all_pallets_with_system );
+		pub type AllPalletsWithoutSystem = ( #all_pallets_without_system );
+
+		/// All pallets included in the runtime as a nested tuple of types in reversed order.
```
```suggestion
		/// All pallets included in the runtime as a nested tuple of types in reversed order, with System pallet last.
```

shawntabrizi commented:
Does `construct_runtime!` require System is first?

thiolliere commented:
> Does `construct_runtime!` require System is first?

no, it is allowed to have it not first

shawntabrizi commented:
> > Does `construct_runtime!` require System is first?
> 
> no, it is allowed to have it not first

Then this should also be a note in the PR IMO, that ordering is relevant even more now

thiolliere commented:
so I added `AllPalletsReversedWithSystemFirst` which allows not to break order at all. (except pre_upgrade/post_upgrade but I think they are minor)

kianenigma reviewed (commented):
The code looks good but abstaining from approving for now since I think we can do a better job at providing an upgrade guide for teams, to do either of

1. use `AllPalletsReversedWithSystem`, which would mean they are sticking to their old logic, with no change to `construct_runtime`

2. Use the new `AllPalletsWithSystem`, but this would mean that they might need to reorder some stuff in `construct_runtime`.

--- 

I also wonder if we could come up with a test scenario to make sure this order never changes. I can imagine something like a ui test, where a representation of the order of `AllPallets` (in any form) is stored as a file and then checked per runtime.

thiolliere commented:
> The code looks good but abstaining from approving for now since I think we can do a better job at providing an upgrade guide for teams, to do either of
> 
>     1. use `AllPalletsReversedWithSystem`, which would mean they are sticking to their old logic, with no change to `construct_runtime`
> 
>     2. Use the new `AllPalletsWithSystem`, but this would mean that they might need to reorder some stuff in `construct_runtime`.
> 
> 
> I also wonder if we could come up with a test scenario to make sure this order never changes. I can imagine something like a ui test, where a representation of the order of `AllPallets` (in any form) is stored as a file and then checked per runtime.

Yes I added your suggestion to provide an upgrade guide,
Order of execution of hooks is tested in the test in frame-support-test, thus this PR changes them.
As you suggest I added some test for all the type alias to ensure they don't change in https://github.com/paritytech/substrate/pull/10043/commits/2180da699391960358275dfd454de71a0e342159

apopiak commented on frame/executive/src/lib.rs:
```diff
@@ -415,30 +411,20 @@ where
+			let used_weight = <AllPalletsWithSystem as OnIdle<System::BlockNumber>>::on_idle(
 				block_number,
 				remaining_weight,
-			)
-			.saturating_add(used_weight);
+			);
```
This looks to me like it changes the semantics of `remaining_weight`. Can you give the reasoning for this change?

apopiak commented on frame/support/procedural/src/construct_runtime/mod.rs:
```diff
@@ -303,35 +303,82 @@ fn decl_all_pallets<'a>(
+			Note that the type `frame_executive::Executive` expects one of `AllPalletsWithSystem` \
+			, `AllPalletsWithSystemReversed`, `AllPalletsReversedWithSystemFirst`. More details in \
+			https://github.com/paritytech/substrate/pull/10043")]
+		pub type AllPallets = AllPalletsWithSystem;
```
We should probably add a note that the previous behavior/structure is available with `AllPalletsWithoutSystemReversed` (IIUC)

thiolliere replied:
yes but considering the usage in executive is changing at the same time, I prefer not to say that people can blindly use `AllPalletsWithoutSystemReversed` when upgrading. Because in the case of executive it needs more care.
I prefered to let the user understand as much as possible to make the good decision when replacing the usage in his codebase.

apopiak replied:
I didn't mean it as a recommendation to blindly use that. I guess I was just pointing out that I would like to have some the excellent info in your PR description in the code comments as well.

thiolliere commented:
bot merge

thiolliere commented:
bot merge

shawntabrizi commented:
what happened with this one?

KiChjang commented:
bot merge

KiChjang commented:
bot merge

KiChjang commented:
bot merge
